packages/sensors/android/src/main/java/io/flutter/plugins/sensors/SensorsPlugin.java

```diff
@@ -6,71 +6,63 @@
 import android.content.Context;
 import android.hardware.Sensor;
-import android.hardware.SensorEvent;
-import android.hardware.SensorEventListener;
 import android.hardware.SensorManager;
+import io.flutter.embedding.engine.plugins.FlutterPlugin;
+import io.flutter.plugin.common.BinaryMessenger;
 import io.flutter.plugin.common.EventChannel;
 import io.flutter.plugin.common.PluginRegistry.Registrar;
 
 /** SensorsPlugin */
-public class SensorsPlugin implements EventChannel.StreamHandler {
+public class SensorsPlugin implements FlutterPlugin {
   private static final String ACCELEROMETER_CHANNEL_NAME =
       "plugins.flutter.io/sensors/accelerometer";
   private static final String GYROSCOPE_CHANNEL_NAME = "plugins.flutter.io/sensors/gyroscope";
   private static final String USER_ACCELEROMETER_CHANNEL_NAME =
       "plugins.flutter.io/sensors/user_accel";
 
+  private EventChannel accelerometerChannel;
+  private EventChannel userAccelChannel;
+  private EventChannel gyroscopeChannel;
+
   /** Plugin registration. */
   public static void registerWith(Registrar registrar) {
-    final EventChannel accelerometerChannel =
-        new EventChannel(registrar.messenger(), ACCELEROMETER_CHANNEL_NAME);
-    accelerometerChannel.setStreamHandler(
-        new SensorsPlugin(registrar.context(), Sensor.TYPE_ACCELEROMETER));
-
-    final EventChannel userAccelChannel =
-        new EventChannel(registrar.messenger(), USER_ACCELEROMETER_CHANNEL_NAME);
-    userAccelChannel.setStreamHandler(
-        new SensorsPlugin(registrar.context(), Sensor.TYPE_LINEAR_ACCELERATION));
-
-    final EventChannel gyroscopeChannel =
-        new EventChannel(registrar.messenger(), GYROSCOPE_CHANNEL_NAME);
-    gyroscopeChannel.setStreamHandler(
-        new SensorsPlugin(registrar.context(), Sensor.TYPE_GYROSCOPE));
-  }
-
-  private SensorEventListener sensorEventListener;
-  private final SensorManager sensorManager;
-  private final Sensor sensor;
-
-  private SensorsPlugin(Context context, int sensorType) {
-    sensorManager = (SensorManager) context.getSystemService(context.SENSOR_SERVICE);
-    sensor = sensorManager.getDefaultSensor(sensorType);
+    SensorsPlugin plugin = new SensorsPlugin();
+    plugin.setupEventChannels(registrar.context(), registrar.messenger());
   }
 
   @Override
-  public void onListen(Object arguments, EventChannel.EventSink events) {
-    sensorEventListener = createSensorEventListener(events);
-    sensorManager.registerListener(sensorEventListener, sensor, sensorManager.SENSOR_DELAY_NORMAL);
+  public void onAttachedToEngine(FlutterPluginBinding binding) {
+    final Context context = binding.getApplicationContext();
+    setupEventChannels(context, binding.getFlutterEngine().getDartExecutor());
   }
 
   @Override
-  public void onCancel(Object arguments) {
-    sensorManager.unregisterListener(sensorEventListener);
+  public void onDetachedFromEngine(FlutterPluginBinding binding) {
+    accelerometerChannel.setStreamHandler(null);
```

Comment:
Recommendation: If you've defined a `setupEventChannels()` that contains setup behavior, consider also defining a `teardownEventChannels()` that does the opposite, rather than directly implementing that behavior in `onDetachedFromEngine()`.